### PR TITLE
Have clients retry the connection if they fail due to "duplicate UID"

### PR DESCRIPTION
**Pull request by ChristianReese**
When a client reboots, and tries to reconnect to the broker, the old connection may still be there, in which case the new connection may fail due to the "duplicate UID" error. In this case, the client should retry, since it will most likely succeed when it does, as the old connection will have been cleaned up. The downside to this is that if it is a user-fixable configuration error of static UIDs, then this will just retry indefinitely.

### CODE REVIEW

**Comment by samkearney** [comment]
We may ultimately want to implement this logic, but this change hides the real bug, which is that there is a timing issue in the broker's handling of dynamic UIDs. We know this because the issue is not happening after a client reboot, but after connection termination due to no heartbeat, when the broker should know that the old connection is closed, but it doesn't update the status of the client's CID. Not implementing retries here allowed us to detect that bug, which would have gone undetected for much longer if retries had been enabled.

**Comment by nwagner** [comment]
We have issues/SCRs open for this, right?

There may be something about the client that may be stopping a graceful close with the broker, but ultimately if the network has problems I want the reconnect to be a better experience.  This was already a problem with some minimal testing.

**Comment by samkearney** [comment]
@nwagner as I mentioned before, this problem that occurred with minimal testing is not due to network problems, but is actually a broker bug which has not yet been fixed. Not having retries allowed us to find the bug. The alternative was that we implemented retries in this condition from the beginning, the broker bug was not found, and it resulted in a much nastier situation later, after release.

**Comment by nwagner** [comment]
Understood.  We can either fix the broker bug before we give this to test, or we can change the code back when we're fixing the bug.  Given what we saw on Friday's minimal testing with the test hardware, I don't feel comfortable giving this to test without a retry in place. We should also look at the client code to see if it is as gracefully dropping the connection to the broker as it can.

**Comment by samkearney** [comment]
Another issue with this change is that a client can get a DUPLICATE_UID message when it is using a static UID, indicating that another client has the same static UID. This is a fatal problem and should not be retried. This is why I included logic for not retrying under certain conditions.

**Comment by nwagner** [comment]
@samkearney , this is the reason we had to add the change.  The client didn't cleanly disconnect (which will certainly happen in the real world), and then a restart of the test hardware failed to connect because the broker still had old session around.  Which would fail any of test's usual disconnect/reconnect/reboot sort of tests..

**Comment by samkearney** [comment]
@nwagner That was not my understanding of the test procedure. My understanding was:

* The client closed the TCP connection cleanly because it received no heartbeat message within the timeout.
* The client retried the connection, and was rejected by the broker because the broker was still tracking the client's CID.
* This is a broker bug, because the broker should have mark the CID as disconnected when the TCP connection was closed.

**Comment by nwagner** [comment]
Well, maybe the misunderstanding is mine, then!  @ChristianReese , care to comment?

**Comment by samkearney** [comment]
@nwagner Your scenario is valid though, even if it didn't occur in the wild yet... Looking at rdmnet/core/client.c, we should probably also retry in a "capacity exceeded" scenario too, since the standard establishes a backoff timer for retries. Socket failure might be the only valid reason not to retry.

**Comment by samkearney** [comment]
Wait sorry, I misunderstood the logic of that function. There are still quite a few connect status codes that will result in not retrying, including "Invalid Client Entry", "Scope Mismatch" and "Invalid Uid". These represent config errors or bugs, so they should still be in the "no retry" category.

File: include/rdmnet/client.h [hunk 1]
```diff
@@ -183,7 +183,7 @@ typedef struct RdmnetClientConnectFailedInfo
    * that the connection failed for a reason that usually must be corrected by a user or
    * application developer. Some possible reasons for this to be false include:
    * - The wrong scope was specified for a statically-configured broker
-   * - A static UID was given that was invalid or duplicate with another UID in the system
+   * - A static UID was given that was invalid
```
Comment (samkearney) [inline include/rdmnet/client.h:186]: Thanks for remembering to update the doc comments.

**Comment by ChristianReese** [comment]
My apologies for the delay in responding - I've been in a meeting. 

The reproducer Sam describes did occur. I haven't been able to reproduce it consistently yet. I agree this is a problem in the broker. I think the broker not sending heartbeats also ties into the defect, since I haven't been seeing random drops of heartbeats in Wireshark, much less three in a row.

However, there is another reproducer as well, where we reboot the Mk2 and then it can't reconnect because of the duplicate UID issue. It seems the issue with that is that it takes a bit of time for the broker to clean up the old connection. I wonder if that's just a quality of the network or if the Mk2 isn't doing something for disconnection. When I set up an example app network on my machine and close the example device, I see it gets removed immediately. That may just be because I'm running it locally, vs. the network lab setup may be introducing some delay.

**Comment by samkearney** [comment]
@ChristianReese to answer your question about why the broker doesn't clean up the old connection: you cannot guarantee a clean disconnect from an embedded device. It's possible that the Mk2 gateway doesn't even bother to gracefully disconnect when rebooting, because it could just as easily be rebooted by someone pulling the power and reconnecting it. It can't reasonably be expected to gracefully close its connection in that circumstance. TCP historically has a very long keep-alive timer, if any, so it can take some time to detect that the connection is unhealthy in this case.

For this reason, the RDMnet standard specifies the transmission of heartbeats and a heartbeat timeout, so that the time to detect an unhealthy connection is capped to 45 seconds. Even then though, as you know, a gateway can easily be rebooted and reconnect within that timeout. So in that case we run into the unavoidable situation of the broker getting a connect request from a client that it thinks is already connected, and we must handle that situation.

**Comment by ChristianReese** [comment]
@samkearney It sounds like this should be handled on the broker in any of these cases. In addition to unplugging, this is able to happen in a normal reboot as well, in which case I think the Mk2 is gracefully disconnecting, but there's still a delay because of the network (because sometimes the connection does clean up before the reboot, and before the 45 seconds, but still with a delay, suggesting there is a graceful disconnect). But the delay apparently can be longer than the reboot, resulting in the error.

The duplicate UID error appears to come from the code where a dynamic UID is being requested, but it already sees a registration on the same CID, and gives the "duplicate UID" error (though "duplicate CID" describes it more closely). Should the fix, then, be to not generate the error, but seamlessly switch to the new connection and clean up the old?

**Comment by samkearney** [comment]
@ChristianReese I disagree that it should be handled on the broker. The broker has no way of knowing if a new connection with a duplicate identifier is due to a timing issue of the type we've described here, or an actual configuration error. This is more obvious when you consider the case of a client with a static UID. The standard specifies that a broker _must_ reject a connection request if it detects a duplicate UID:

> If the Client is an RPT Client with a Static UID and its UID matches the UID of any RPT Client in the Broker's Connected Client List, the Broker shall send a Connect Reply Broker PDU with Connection Code CONNECT_DUPLICATE_UID.

And this is necessary because there might actually be clients with duplicate UIDs in the wild, and there must be a way to detect this error condition.

I think it's OK to resolve this by having the client retry in this scenario. The old connection will clean itself up within 45 seconds, at which point the client will be able to connect again.

There is still a problem with the broker's handling of Dynamic UID registrations, though. I believe it should mark a CID/dynamic UID pair as disconnected as soon as the connection is closed, but it waits until the connection is cleaned up (BrokerCore::DestroyMarkedClients()), which happens less frequently on a different thread, thus dramatically increasing the likelihood that there will be a conflict even on a graceful connection.

**Comment by ChristianReese** [comment]
@samkearney I suppose it's also technically possible for two devices to be misconfigured to the same CID, however unlikely. So the same principle applies - this condition should be detected.

I'll look into that registration issue. Beyond that, the only other thing I can think of that may need fixing in the broker is the case where the broker does not send heartbeats, and still doesn't clean up the connection even though 45 seconds have elapsed. I might play around with this more and see if I can figure out what's going on.

**Comment by ChristianReese** [comment]
I just realized that the broker wouldn't clean up the connection in the case where it doesn't send heartbeats - even then, it was probably still getting heartbeats from the gateway, so the connection would be kept alive without going near the 45 second limit. The Mk2, on the other hand, makes it to 45 seconds and reboots itself, resulting in the exact same situation as if we rebooted the Mk2 manually - the graceful disconnect arrives too late, the broker sees the old connection, and replies with Duplicate UID. That still leaves the question of why the broker stopped sending heartbeats - but merging this now shouldn't really hinder the debugging of that, since if it happens, it will be logged on both sides.

Given the retries will make this testable, I'll make an SCR to monitor/investigate the heartbeat issue, but not a very high priority.

I'll also go ahead and adjust the registration update code Sam mentioned. After that I will merge and close.

**Pull request merged**